File: scripts/optimize-images.js
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

const directoryPath = path.join(__dirname, 'public', 'images');

const optimizeImages = async (dir) => {
    fs.readdir(dir, { withFileTypes: true }, async (err, files) => {
        if (err) {
            console.error("Error reading directory:", err);
            return;
        }

        for (const file of files) {
            const fullPath = path.join(dir, file.name);

            if (file.isDirectory()) {
                await optimizeImages(fullPath);
            } else if (/\.(jpg|jpeg|png)$/i.test(file.name)) {
                try {
                    const tempPath = fullPath + '.tmp';
                    const stats = fs.statSync(fullPath);
                    const originalSizeMB = (stats.size / (1024 * 1024)).toFixed(2);

                    let sharpInstance = sharp(fullPath);

                    if (/\.(jpg|jpeg)$/i.test(file.name)) {
                        sharpInstance = sharpInstance.jpeg({ quality: 60, progressive: true });
                    } else if (/\.png$/i.test(file.name)) {
                        sharpInstance = sharpInstance.png({ quality: 60, compressionLevel: 9 });
                    }

                    // Resize if width is larger than 1920px (full HD)
                    const metadata = await sharpInstance.metadata();
                    if (metadata.width > 1920) {
                        sharpInstance = sharpInstance.resize({ width: 1920, withoutEnlargement: true });
                    }

                    await sharpInstance.toFile(tempPath);

                    fs.renameSync(tempPath, fullPath);

                    const newStats = fs.statSync(fullPath);
                    const newSizeMB = (newStats.size / (1024 * 1024)).toFixed(2);

                    if (originalSizeMB !== newSizeMB) {
                        console.log(`Optimized ${file.name}: ${originalSizeMB}MB -> ${newSizeMB}MB`);
                    }

                } catch (error) {
                    console.error(`Error optimizing ${file.name}:`, error);
                }
            }
        }
    });
};

console.log("Starting image optimization...");
optimizeImages(directoryPath);

File: scripts/optimize-images.py
import os
import glob
from PIL import Image

def optimize_images(directory):
    # Find all jpegs and pngs
    extensions = ('*.jpg', '*.jpeg', '*.png')
    files = []
    for ext in extensions:
        files.extend(glob.glob(os.path.join(directory, '**', ext), recursive=True))

    for file in files:
        try:
            original_size = os.path.getsize(file)
            original_size_mb = original_size / (1024 * 1024)
            
            # Skip if already very small (< 100KB)
            if original_size < 100 * 1024:
                continue

            with Image.open(file) as img:
                # Convert to RGB if RGBA (needed for JPEG saving)
                if img.mode in ("RGBA", "P"):
                    img = img.convert("RGB")
                    
                # Resize if width > 1920
                if img.width > 1920:
                    new_height = int((1920 / img.width) * img.height)
                    img = img.resize((1920, new_height), Image.Resampling.LANCZOS)

                # Save over original with optimized settings
                if file.lower().endswith('.png'):
                    img.save(file, optimize=True)
                else:
                    img.save(file, "JPEG", quality=60, optimize=True, progressive=True)

            new_size = os.path.getsize(file)
            new_size_mb = new_size / (1024 * 1024)
            
            if new_size < original_size:
                print(f"Optimized {os.path.basename(file)}: {original_size_mb:.2f}MB -> {new_size_mb:.2f}MB")
        except Exception as e:
            print(f"Error optimizing {file}: {e}")

if __name__ == "__main__":
    base_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "public", "images")
    if os.path.exists(base_dir):
        print("Starting image optimization with Python Pillow...")
        optimize_images(base_dir)
        print("Done!")
    else:
        print(f"Directory not found: {base_dir}")
